### PR TITLE
Make relase_notes.rb syntax-aware

### PR DESCRIPTION
**Pull request by lzap**
Untested, @melcorr let me know if it works.

### CODE REVIEW

File: scripts/release_notes.rb [hunk 1]
```diff
@@ -10,14 +10,19 @@
 
 URL = 'https://projects.theforeman.org'
 
-unless ARGV.size == 2
-  puts "Usage: #{$0} PROJECTS RELEASE_NAME"
+unless ARGV.size <= 2
```
Comment (ekohl) [inline scripts/release_notes.rb]: I think this isn't correct. Shouldn't it be:
```suggestion
if ARGV.size < 2
```

**Review by ekohl — COMMENTED** [review]
Overall :+1: but I think the conditional isn't correct.

File: scripts/release_notes.rb [hunk 2]
```diff
@@ -10,14 +10,19 @@
...
+  puts "Usage: #{$0} PROJECTS RELEASE_NAME [markdown|asciidoc]"
   puts "Example: #{$0} foreman 1.18.0"
   exit 1
 end
 
 @project_names = ARGV[0].split(',')
 @current_release_name = ARGV[1]
+@output = ARGV[2] || "markdown"
+unless ["markdown", "asciidoc"].includes(@output)
```
Comment (upadhyeammit) [inline scripts/release_notes.rb]: The method should be `include?` instead of `includes` ?

File: scripts/release_notes.rb [hunk 3]
```diff
@@ -78,7 +83,11 @@ def gather_issues(offset = 0)
   puts "#### #{category}"
 
   issues.each do |issue|
-    puts "* #{issue['subject'].gsub('`','\\\`').gsub('<','&lt;').gsub('>','&gt;').gsub('*','\\\*')} \\- [##{issue['id']}](#{URL}/issues/#{issue['id']})"
+    if @output == "markdown"
+      puts "* #{issue['subject'].gsub('`','\\\`').gsub('<','&lt;').gsub('>','&gt;').gsub('*','\\\*')} \\- [##{issue['id']}](#{URL}/issues/#{issue['id']})"
+    else
+      puts "* #{issue['subject'].gsub('`','\\\`').gsub('<','&lt;').gsub('>','&gt;').gsub('*','\\\*')} \\- #{URL}/issues/#{issue['id']}[##{issue['id']}]"
```
Comment (upadhyeammit) [inline scripts/release_notes.rb]: I think we dont need escape char before `-` as I see `\-` is getting rendered on: https://asciidoclive.com/edit/scratch/1 with,
~~~
* double Puppet deprecation notice on host creation page \- https://projects.theforeman.org/issues/32493[#32493]
~~~
```suggestion
      puts "* #{issue['subject'].gsub('`','\\\`').gsub('<','&lt;').gsub('>','&gt;').gsub('*','\\\*')} - #{URL}/issues/#{issue['id']}[##{issue['id']}]"
```

File: scripts/release_notes.rb [hunk 4]
```diff
@@ -10,14 +10,19 @@
 
 URL = 'https://projects.theforeman.org'
 
-unless ARGV.size == 2
-  puts "Usage: #{$0} PROJECTS RELEASE_NAME"
+unless ARGV.size <= 2
+  puts "Usage: #{$0} PROJECTS RELEASE_NAME [markdown|asciidoc]"
```
Comment (upadhyeammit) [inline scripts/release_notes.rb:14]: Should we print what is the default output mode or not required?

File: scripts/release_notes.rb (hunk 2 above)
Comment (lzap) [reply scripts/release_notes.rb]: Yeah sorry just tested this. Fixing.

File: scripts/release_notes.rb (hunk 1 above)
Comment (lzap) [reply scripts/release_notes.rb]: Sure.

File: scripts/release_notes.rb (hunk 3 above)
Comment (lzap) [reply scripts/release_notes.rb]: Thanks I was rushing out quick and it was completely untested.

File: scripts/release_notes.rb (hunk 4 above)
Comment (lzap) [reply scripts/release_notes.rb:14]: Yup.

**Comment by lzap** [comment]
Resolved all bugs and comments, here is an example:

```
$ scripts/release_notes.rb foreman 2.5.1
scripts/release_notes.rb:29: warning: URI.escape is obsolete
Copy and paste the following into the release notes section of the relevant manual:

----------[Start of notes]----------

### Release notes for 2.5.1
scripts/release_notes.rb:29: warning: URI.escape is obsolete
#### Foreman - Compute resources - VMware
* Can not modify associated hosts if connection to vCenter is down \- [#31307](https://projects.theforeman.org/issues/31307)

#### Foreman - DB migrations
* Migrations fail with 'could not create unique index "index_lookup_values_on_lookup_key_id_and_match"' error \- [#32673](https://projects.theforeman.org/issues/32673)

#### Foreman - Database
* PG::AmbiguousColumn when listing domains \- [#32641](https://projects.theforeman.org/issues/32641)

#### Foreman - Host creation
* Confirming interface modal by enter causes whole Host form to get confirmed \- [#32397](https://projects.theforeman.org/issues/32397)

#### Foreman - Host registration
* Normally Inherited fields are not inherited upon host registreation \- [#32457](https://projects.theforeman.org/issues/32457)
* Add support for AlmaLinux 8 as a client \- [#31787](https://projects.theforeman.org/issues/31787)

#### Foreman - Inventory
* Host api request fails when thin=true and per_page=all \- [#32776](https://projects.theforeman.org/issues/32776)
* API - ISE when using invalid status type \- [#32536](https://projects.theforeman.org/issues/32536)
* Add Rocky Linux support to Foreman \- [#32516](https://projects.theforeman.org/issues/32516)

#### Foreman - Search
* Editing an existing bookmark should respect the current visibility \- [#32662](https://projects.theforeman.org/issues/32662)
* Allow searching hosts by attributes of the reported data facet \- [#32140](https://projects.theforeman.org/issues/32140)

#### Foreman - Settings
* CVE-2021-3584: Remote code execution through Sendmail configuration \- [#32753](https://projects.theforeman.org/issues/32753)

#### Foreman - Templates
* Avoid echo statements without quotes in Grub \- [#32681](https://projects.theforeman.org/issues/32681)
* kickstart_kernel_options doesn't handle properly tagged vlan over bond device \- [#32497](https://projects.theforeman.org/issues/32497)

#### Foreman - Unattended installations
* puppet.conf template do not render after introduction of puppet host params \- [#32772](https://projects.theforeman.org/issues/32772)

#### Foreman - Users, Roles and Permissions
* Non-admin user can not revoke user access token, attempt ends with 404  \- [#32826](https://projects.theforeman.org/issues/32826)
* 50/50 chance to create role filter with non-admin user and enough permissions \- [#30394](https://projects.theforeman.org/issues/30394)

#### Foreman - Web Interface
* Update all freenode references to point to https://theforeman.org/support.html#IRClivechat or libera.chat \- [#32809](https://projects.theforeman.org/issues/32809)

#### Installer - Foreman modules
* Additional RPMs in /pub directory break latest symlink \- [#32700](https://projects.theforeman.org/issues/32700)
* remove md5 from allowed content types for pulpcore \- [#32553](https://projects.theforeman.org/issues/32553)

#### Packaging
* foreman 2.5 /etc/cron.d/foreman broken \- [#32859](https://projects.theforeman.org/issues/32859)

#### Packaging - RPMs
* katello-change-hostname assumes answers file location \- [#32743](https://projects.theforeman.org/issues/32743)
* katello-change-hostname should handle --no-enable-puppet properly \- [#32725](https://projects.theforeman.org/issues/32725)

*A full list of changes in 2.5.1 is available via [Redmine](https://projects.theforeman.org/issues?set_filter=1&sort=id%3Adesc&status_id=closed&f[]=cf_12&op[cf_12]=%3D&v[cf_12][]=1409)*

----------[End of notes]----------
If this isn't an RC release, the relevant releases should be closed (not possible via API):
Close the release https://projects.theforeman.org/versions/1409/edit
```

**Comment by lzap** [comment]
And asciidoc:

```
$ scripts/release_notes.rb foreman 2.5.1
scripts/release_notes.rb:29: warning: URI.escape is obsolete
Copy and paste the following into the release notes section of the relevant manual:

----------[Start of notes]----------

### Release notes for 2.5.1
scripts/release_notes.rb:29: warning: URI.escape is obsolete
#### Foreman - Compute resources - VMware
* Can not modify associated hosts if connection to vCenter is down \- [#31307](https://projects.theforeman.org/issues/31307)

#### Foreman - DB migrations
* Migrations fail with 'could not create unique index "index_lookup_values_on_lookup_key_id_and_match"' error \- [#32673](https://projects.theforeman.org/issues/32673)

#### Foreman - Database
* PG::AmbiguousColumn when listing domains \- [#32641](https://projects.theforeman.org/issues/32641)

#### Foreman - Host creation
* Confirming interface modal by enter causes whole Host form to get confirmed \- [#32397](https://projects.theforeman.org/issues/32397)

#### Foreman - Host registration
* Normally Inherited fields are not inherited upon host registreation \- [#32457](https://projects.theforeman.org/issues/32457)
* Add support for AlmaLinux 8 as a client \- [#31787](https://projects.theforeman.org/issues/31787)

#### Foreman - Inventory
* Host api request fails when thin=true and per_page=all \- [#32776](https://projects.theforeman.org/issues/32776)
* API - ISE when using invalid status type \- [#32536](https://projects.theforeman.org/issues/32536)
* Add Rocky Linux support to Foreman \- [#32516](https://projects.theforeman.org/issues/32516)

#### Foreman - Search
* Editing an existing bookmark should respect the current visibility \- [#32662](https://projects.theforeman.org/issues/32662)
* Allow searching hosts by attributes of the reported data facet \- [#32140](https://projects.theforeman.org/issues/32140)

#### Foreman - Settings
* CVE-2021-3584: Remote code execution through Sendmail configuration \- [#32753](https://projects.theforeman.org/issues/32753)

#### Foreman - Templates
* Avoid echo statements without quotes in Grub \- [#32681](https://projects.theforeman.org/issues/32681)
* kickstart_kernel_options doesn't handle properly tagged vlan over bond device \- [#32497](https://projects.theforeman.org/issues/32497)

#### Foreman - Unattended installations
* puppet.conf template do not render after introduction of puppet host params \- [#32772](https://projects.theforeman.org/issues/32772)

#### Foreman - Users, Roles and Permissions
* Non-admin user can not revoke user access token, attempt ends with 404  \- [#32826](https://projects.theforeman.org/issues/32826)
* 50/50 chance to create role filter with non-admin user and enough permissions \- [#30394](https://projects.theforeman.org/issues/30394)

#### Foreman - Web Interface
* Update all freenode references to point to https://theforeman.org/support.html#IRClivechat or libera.chat \- [#32809](https://projects.theforeman.org/issues/32809)

#### Installer - Foreman modules
* Additional RPMs in /pub directory break latest symlink \- [#32700](https://projects.theforeman.org/issues/32700)
* remove md5 from allowed content types for pulpcore \- [#32553](https://projects.theforeman.org/issues/32553)

#### Packaging
* foreman 2.5 /etc/cron.d/foreman broken \- [#32859](https://projects.theforeman.org/issues/32859)

#### Packaging - RPMs
* katello-change-hostname assumes answers file location \- [#32743](https://projects.theforeman.org/issues/32743)
* katello-change-hostname should handle --no-enable-puppet properly \- [#32725](https://projects.theforeman.org/issues/32725)

*A full list of changes in 2.5.1 is available via [Redmine](https://projects.theforeman.org/issues?set_filter=1&sort=id%3Adesc&status_id=closed&f[]=cf_12&op[cf_12]=%3D&v[cf_12][]=1409)*

----------[End of notes]----------
If this isn't an RC release, the relevant releases should be closed (not possible via API):
Close the release https://projects.theforeman.org/versions/1409/edit
[lzap@nuc theforeman.org]$ v scripts/release_notes.rb
[lzap@nuc theforeman.org]$ scripts/release_notes.rb foreman 2.5.1 asciidoc
scripts/release_notes.rb:29: warning: URI.escape is obsolete
Copy and paste the following into the release notes section of the relevant manual:

----------[Start of notes]----------

### Release notes for 2.5.1
scripts/release_notes.rb:29: warning: URI.escape is obsolete
#### Foreman - Compute resources - VMware
* Can not modify associated hosts if connection to vCenter is down - https://projects.theforeman.org/issues/31307[#31307]

#### Foreman - DB migrations
* Migrations fail with 'could not create unique index "index_lookup_values_on_lookup_key_id_and_match"' error - https://projects.theforeman.org/issues/32673[#32673]

#### Foreman - Database
* PG::AmbiguousColumn when listing domains - https://projects.theforeman.org/issues/32641[#32641]

#### Foreman - Host creation
* Confirming interface modal by enter causes whole Host form to get confirmed - https://projects.theforeman.org/issues/32397[#32397]

#### Foreman - Host registration
* Normally Inherited fields are not inherited upon host registreation - https://projects.theforeman.org/issues/32457[#32457]
* Add support for AlmaLinux 8 as a client - https://projects.theforeman.org/issues/31787[#31787]

#### Foreman - Inventory
* Host api request fails when thin=true and per_page=all - https://projects.theforeman.org/issues/32776[#32776]
* API - ISE when using invalid status type - https://projects.theforeman.org/issues/32536[#32536]
* Add Rocky Linux support to Foreman - https://projects.theforeman.org/issues/32516[#32516]

#### Foreman - Search
* Editing an existing bookmark should respect the current visibility - https://projects.theforeman.org/issues/32662[#32662]
* Allow searching hosts by attributes of the reported data facet - https://projects.theforeman.org/issues/32140[#32140]

#### Foreman - Settings
* CVE-2021-3584: Remote code execution through Sendmail configuration - https://projects.theforeman.org/issues/32753[#32753]

#### Foreman - Templates
* Avoid echo statements without quotes in Grub - https://projects.theforeman.org/issues/32681[#32681]
* kickstart_kernel_options doesn't handle properly tagged vlan over bond device - https://projects.theforeman.org/issues/32497[#32497]

#### Foreman - Unattended installations
* puppet.conf template do not render after introduction of puppet host params - https://projects.theforeman.org/issues/32772[#32772]

#### Foreman - Users, Roles and Permissions
* Non-admin user can not revoke user access token, attempt ends with 404  - https://projects.theforeman.org/issues/32826[#32826]
* 50/50 chance to create role filter with non-admin user and enough permissions - https://projects.theforeman.org/issues/30394[#30394]

#### Foreman - Web Interface
* Update all freenode references to point to https://theforeman.org/support.html#IRClivechat or libera.chat - https://projects.theforeman.org/issues/32809[#32809]

#### Installer - Foreman modules
* Additional RPMs in /pub directory break latest symlink - https://projects.theforeman.org/issues/32700[#32700]
* remove md5 from allowed content types for pulpcore - https://projects.theforeman.org/issues/32553[#32553]

#### Packaging
* foreman 2.5 /etc/cron.d/foreman broken - https://projects.theforeman.org/issues/32859[#32859]

#### Packaging - RPMs
* katello-change-hostname assumes answers file location - https://projects.theforeman.org/issues/32743[#32743]
* katello-change-hostname should handle --no-enable-puppet properly - https://projects.theforeman.org/issues/32725[#32725]

*A full list of changes in 2.5.1 is available via [Redmine](https://projects.theforeman.org/issues?set_filter=1&sort=id%3Adesc&status_id=closed&f[]=cf_12&op[cf_12]=%3D&v[cf_12][]=1409)*

----------[End of notes]----------
If this isn't an RC release, the relevant releases should be closed (not possible via API):
Close the release https://projects.theforeman.org/versions/1409/edit
```

**Review by upadhyeammit — APPROVED** [review]
LGTM!

**Comment by upadhyeammit** [comment]
@ekohl any more suggestions from your side?

**Review by ekohl — CHANGES_REQUESTED** [review]
https://github.com/theforeman/theforeman.org/pull/1880/files#r693967090 is still open:
```console
$ ./scripts/release_notes.rb 
./scripts/release_notes.rb:19:in `<main>': undefined method `split' for nil:NilClass (NoMethodError)
$ ./scripts/release_notes.rb foreman 3.0.0 asciidoc
Usage: ./scripts/release_notes.rb PROJECTS RELEASE_NAME [markdown|asciidoc]
Example: ./scripts/release_notes.rb foreman 1.18.0
```

Note that https://github.com/theforeman/theforeman.org/pull/1880#issuecomment-903960593 didn't actually show asciidoc

**Comment by upadhyeammit** [comment]
@ekohl and @lzap as the plan is to move docs to new site then shouldn't we move the tools required for docs to new repo? I think we should keep this unchanged in this repo and actually have similar or new script in https://github.com/theforeman/foreman-documentation?

**Comment by ekohl** [comment]
It does make sense and in time it should certainly move over. In the short term I'd be ok with changing it here but I'd be just as happy with a script that creates a proper guide in foreman-documentation.

**Review by ekohl — CHANGES_REQUESTED** [review]
The full list at the end also needs to be an asciidoc link.

**Comment by lzap** [comment]
I forgot to push, pushed now.

> Note that [#1880 (comment)](https://github.com/theforeman/theforeman.org/pull/1880#issuecomment-903960593) didn't actually show asciidoc

It shows both, just to make sure it works. Scroll down a bit to see asciidoc version.

**Comment by ekohl** [comment]
I've taken this as a basis and implemented it in https://github.com/theforeman/foreman-documentation/pull/698 (taking @upadhyeammit's suggestion). I'm proposing we close this PR. Having the script live in foreman-documentation means we'll also get it right with branching etc. See https://github.com/theforeman/foreman-documentation/pull/698 for the generated result.

**Comment by lzap** [comment]
Sure, thanks for looking into it.

**Pull request closed without merging**